### PR TITLE
stm32h7: fix config conflict

### PR DESCRIPTION
**Pull request by 0528Mike**
## Summary
fix config conflict
#if defined(CONFIG_NSH_LIBRARY) && !defined(CONFIG_BOARDCTL)
## Impact
No Impact
## Testing
None

### CODE REVIEW

**Review by raiden00pl — CHANGES_REQUESTED** [review]
What is the motivation for this change? GPIO encoding is intended to be handled by logic in `arch/arm/src/stm32h7/stm32_gpio.c`. 
With this change, some of the pin configuration must be handled by the board logic, some by the arch logic. This introduces unnecessary complications. In my opinion this is a bad change.

**Comment by 0528Mike** [comment]
defined(CONFIG_NSH_LIBRARY) and !defined(CONFIG_BOARDCTL) is conflict，does this need to be modified?

**Comment by 0528Mike** [comment]
> defined(CONFIG_NSH_LIBRARY) and !defined(CONFIG_BOARDCTL) is conflict，does this need to be modified?

**Comment by raiden00pl** [comment]
> defined(CONFIG_NSH_LIBRARY) and !defined(CONFIG_BOARDCTL) is conflict，does this need to be modified?

this change looks good

**Comment by acassis** [comment]
@0528Mike @raiden00pl I think we could remove CONFIG_NSH_ARCHINIT all together, since CONFIG_BOARD_LATE_INITIALIZE could be used as a generic alternative. What do you think?

**Comment by raiden00pl** [comment]
> @0528Mike @raiden00pl I think we could remove CONFIG_NSH_ARCHINIT all together, since CONFIG_BOARD_LATE_INITIALIZE could be used as a generic alternative. What do you think?

this seems like a good idea, but we have to wonder if this won't break something. This change would be quite big (update apps and all boards with CONFIG_NSH_ARCHINIT). @acassis  maybe you can create an issue about this? so that the problem is more visible and more people can comment.

**Comment by acassis** [comment]
> > @0528Mike @raiden00pl I think we could remove CONFIG_NSH_ARCHINIT all together, since CONFIG_BOARD_LATE_INITIALIZE could be used as a generic alternative. What do you think?
> 
> this seems like a good idea, but we have to wonder if this won't break something. This change would be quite big (update apps and all boards with CONFIG_NSH_ARCHINIT). @acassis maybe you can create an issue about this? so that the problem is more visible and more people can comment.

I think I already did it, let me check

**Comment by acassis** [comment]
Actually I only commented about it in the issue #10821

**Comment by xiaoxiang781216** [comment]
@raiden00pl can we merge this patch now?

**Comment by raiden00pl** [comment]
@xiaoxiang781216 https://github.com/apache/nuttx/pull/11306/files#r1413026144 this hasn't been resolved yet

File: boards/arm/stm32h7/nucleo-h743zi/src/nucleo-h743zi.h [hunk 1]
```diff
@@ -236,9 +236,6 @@
  *   CONFIG_BOARD_LATE_INITIALIZE=y :
  *     Called from board_late_initialize().
  *
- *   CONFIG_BOARD_LATE_INITIALIZE=y && CONFIG_BOARDCTL=y :
- *     Called from the NSH library
- *
```
Comment (raiden00pl) [inline boards/arm/stm32h7/nucleo-h743zi/src/nucleo-h743zi.h:241]:
```
 *   CONFIG_BOARD_LATE_INITIALIZE=n && CONFIG_BOARDCTL=y &&
 *   CONFIG_NSH_ARCHINIT:
 *     Called from the NSH library
 ```

File: boards/arm/stm32h7/stm32h747i-disco/src/stm32h747i-disco.h [hunk 2]
```diff
@@ -115,9 +115,6 @@
  *   CONFIG_BOARD_LATE_INITIALIZE=y :
  *     Called from board_late_initialize().
  *
- *   CONFIG_BOARD_LATE_INITIALIZE=y && CONFIG_BOARDCTL=y :
```
Comment (raiden00pl) [inline boards/arm/stm32h7/stm32h747i-disco/src/stm32h747i-disco.h:118]:
```
 *   CONFIG_BOARD_LATE_INITIALIZE=n && CONFIG_BOARDCTL=y &&
 *   CONFIG_NSH_ARCHINIT:
 *     Called from the NSH library
 ```

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**